arangod/Aql/ConstrainedSortExecutor.cpp

```diff
@@ -0,0 +1,257 @@
+#include "Aql/AqlItemBlockShell.h"
+#include "Aql/AqlItemMatrix.h"
+#include "Aql/ExecutionBlockImpl.h"
+#include "Aql/InputAqlItemRow.h"
+#include "Aql/OutputAqlItemRow.h"
+#include "Aql/SortRegister.h"
+#include "Aql/Stats.h"
+
+#include <algorithm>
+
+using namespace arangodb;
+using namespace arangodb::aql;
+
+namespace {
+
+//dirty trick to access private member is standard conform way
+template <typename Tag, typename Tag::type member>
+struct createAccess {
+  friend typename Tag::type get(Tag) { return member; } // ADL function get in namespace
+};
+
+struct RowTag {
+      using type =  size_t OutputAqlItemRow::*;
+      friend type get(RowTag); // finds the get functions via ADL
+};
+
+template struct createAccess<RowTag, &OutputAqlItemRow::_baseIndex>; // in template instantiation not access check is done
+
+void eraseRow(AqlItemBlockShell& shell, size_t row) {
+  arangodb::aql::RegisterId const nrRegs = shell.block().getNrRegs();
+  for (size_t i = 0; i < nrRegs; i++) {
+    shell.block().destroyValue(row, i);
+  }
+}
+
+}  // namespace
+
+/// @brief OurLessThan
+class arangodb::aql::ConstrainedLessThan {
+ public:
+  ConstrainedLessThan(arangodb::transaction::Methods* trx,
+                      std::vector<arangodb::aql::SortRegister>& sortRegisters) noexcept
+      : _trx(trx), _lhsBuffer(nullptr), _rhsBuffer(nullptr), _sortRegisters(sortRegisters) {}
+
+  void setBuffers(arangodb::aql::AqlItemBlock* lhsBuffer,
+                  arangodb::aql::AqlItemBlock* rhsBuffer) {
+    _lhsBuffer = lhsBuffer;
+    _rhsBuffer = rhsBuffer;
+  }
+
+  bool operator()(uint32_t const& a, uint32_t const& b) const {
+    TRI_ASSERT(_lhsBuffer);
+    TRI_ASSERT(_rhsBuffer);
+
+    for (auto const& reg : _sortRegisters) {
+      auto const& lhs = _lhsBuffer->getValueReference(a, reg.reg);
+      auto const& rhs = _rhsBuffer->getValueReference(b, reg.reg);
+
+      int const cmp = arangodb::aql::AqlValue::Compare(_trx, lhs, rhs, true);
+
+      if (cmp < 0) {
+        return reg.asc;
+      } else if (cmp > 0) {
+        return !reg.asc;
+      }
+    }
+
+    return false;
+  }
+
+ private:
+  arangodb::transaction::Methods* _trx;
+  arangodb::aql::AqlItemBlock* _lhsBuffer;
+  arangodb::aql::AqlItemBlock* _rhsBuffer;
+  std::vector<arangodb::aql::SortRegister>& _sortRegisters;
+};  // ConstrainedLessThan
+
+arangodb::Result ConstrainedSortExecutor::pushRow(InputAqlItemRow& input) {
+  LOG_DEVEL << "enter push row";
+  using arangodb::aql::AqlItemBlock;
+  using arangodb::aql::AqlValue;
+  using arangodb::aql::RegisterId;
+
+  size_t dRow = _rowsPushed;
+
+  if (_rowsPushed >= _infos._limit) {
+    // pop an entry first
+    std::pop_heap(_rows.begin(), _rows.end(), *_cmpHeap.get());
+    dRow = _rows.back();
+     eraseRow(*_heapBuffer, dRow);
+    _rows.pop_back();
+  }
+  TRI_ASSERT(dRow < _infos._limit);
+
+  TRI_IF_FAILURE("SortBlock::doSortingInner") {
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+  }
+
+  _heapOutPutRow.*get(RowTag()) = dRow;
```

Comment:
I'm very confused by this line. Could you explain what this is doing and how it works?

Reply:
It can not stay like that! It is a hack to access a private member. `obj.*member_pointer` is the syntax to access a member via a pointer. The get() function, that can only be found via ADL, provides that member pointer. The pointer to that member is acquired by a template instantiation, as there will no access (private / public) check be done.
I want to get the code working, without breaking other code. Then we can take a look with @goedderz how to improve the performance without breaking the encapsulation completely.

Reply:
Okay. So am I correct then in understanding that this is just resetting the internal row position in the buffer block? That's fine if so. It was just very confusing since I had never seen the syntax before. That is indeed a dirty little trick.

Reply:
yes @danielhlarkin that is exactly what is supposed to do, also disapproved the hack, is now fixed.